### ISSUE TITLE
Support for openai deployment quotas

### ISSUE THREAD

**Issue by kliseb** [hashicorp/terraform-provider-azurerm#22135]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Azure introduced a new [quota feature for openai deployments](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/how-to/quota). This request is to add the Tokens-Per-Minute (TPM) quota to `azurerm_cognitive_deployment` resources.

### New or Affected Resource(s)/Data Source(s)

azurerm_cognitive_deployment

### Potential Terraform Configuration

```hcl
resource "azurerm_cognitive_deployment" "deployment" {
   name                 = "deployment_name"
   cognitive_account_id = azurerm_cognitive_account.openai_account.id
   model {
     format  = "OpenAI"
     name    = "text-embedding-ada-002"
     version = "2"
   }

+  tpm = 120000
+
   scale {
     type = "Standard"
   }
 }
```

**Comment by kliseb** [comment]
After manually updating the TPM through Azure AI Studio, terraform is wanting to replace the deployment which would result in our quota being reset. It seems the setting the quota through the UI is silently changing `rai_policy_name`.

```  # module.openai.azurerm_cognitive_deployment.deployment must be replaced
-/+ resource "azurerm_cognitive_deployment" "deployment" {
      ~ id                   = "redacted" -> (known after apply)
        name                 = "gpt-35-turbo"
      - rai_policy_name      = "Microsoft.Default" -> null # forces replacement
        # (1 unchanged attribute hidden)

        # (2 unchanged blocks hidden)
    }
```

**Comment by liuwuliuyun** [comment]
Hi @kliseb, thanks for raising this issue. We do have plans to upgrade this resource. This will take two steps. 1. Upgrade the api version for all congnitive service related resources. 2. add the missing properties in new api version. But if you'd perfer to have the new property right now. I would suggest you could try [Azapi Provider](https://registry.terraform.io/providers/Azure/azapi/latest) in which you could select the latest version of API and use the new properties. In the meamwhile, we will working on the updates.

**Comment by efiedle6** [comment]
@liuwuliuyun is there a tentative timeline on when this upgrade will go live?

**Comment by liuwuliuyun** [comment]
Hi all, I double checked the latest API and [rate limit](https://github.com/Azure/azure-rest-api-specs/blob/89e91a02e7cf20da48bb84c34eda694faed4dc5c/specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2023-05-01/cognitiveservices.json#L3539C9-L3542C11) is a read-only property in the latest rest API. Since AzureRM provider depends on REST API to perform PUT action, we are not able to support this property anytime soon. You could manually change in the Azure Portal and refresh terraform. To avoid conflict menthioned above. You could leave the `rai_policy_name` empty.

**Comment by liuwuliuyun** [comment]
Hi @justinrush , I don't think that is the token limit property. But if anyone wants to try the new feature once the API is released, here is an terraform template using AzAPI

```
provider "azurerm" {
  features {
    
  }
}

resource "azurerm_resource_group" "example" {
  name = "testyunliu"
  location = "East US"
}

resource "azurerm_cognitive_account" "example" {
  name                = "example-ca"
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
  kind                = "OpenAI"
  sku_name            = "S0"
}

resource "azapi_resource" "example" {
  type = "Microsoft.CognitiveServices/accounts/deployments@2023-05-01"
  name = "testdep"
  parent_id = azurerm_cognitive_account.example.id
  body = jsonencode({
    # PUT the rest api body here
  })
}
```

**Comment by Alfonsoaz01** [comment]
Good morning @kliseb,

I do not know if this help. I have found the value that helps you to set up the capacity and the tokens per minute in the deployments in terraform. 

`resource "azurerm_cognitive_deployment" "deployment_gpt" {
 
  cognitive_account_id = azurerm_cognitive_account.openai.id
  name                 = "gpt-35-turbo"


  model {
    format  = "OpenAI"
    name    = "gpt-35-turbo"
    version = "0301"

  }
  **scale {
    type     = "Standard"
    capacity = 60 # Here is the exact value to adapt the tokens per minute or the capacity
  }**

}`

I hope this helps! 
Cheers

**Comment by kliseb** [comment]
Hi @Alfonsoaz01, thank you for the tip. I agree that looks like the correct argument for setting TPM, as the UI also shows TPM as the deployment's "Capacity". Unfortunately azurerm provider limits its value to between 1-10000 and we currently have a capacity of 840000, so this argument's possible values need to be updated.

**Comment by liuwuliuyun** [comment]
Hi everyone, thanks for the updates. I will working on improve this property.

**Comment by miladh** [comment]
Hi, AFAICS the limitation on capacity was fixed and released in version 3.66.0. However, I'm still getting this when the `capacity` is set to a number lager than 10K:

```
...deployment should be at least 1 and no more than 10000. 
```

@kliseb have you been able to set a higher limit with version 3.66.0 and beyond?

**Comment by kliseb** [comment]
Yes, sorry, I meant to come back and update this. I didn't realize initially that the capacity was given in thousands.
